src/Common/tests/Collections/IEnumerableTest.cs

```diff
@@ -0,0 +1,428 @@
+        /// <returns>An instance of the enumerable under test containing the given items.</returns>
+        protected abstract IEnumerable GetEnumerable(object[] items);
+
+        /// <summary>
+        ///     When overridden in a derived class, invalidates any enumerators for the given IEnumerable.
+        /// </summary>
+        /// <param name="enumerable">The <see cref="IEnumerable"/> to invalidate enumerators for.</param>
+        /// <returns>The new set of items in the <see cref="IEnumerable"/></returns>
+        protected abstract object[] InvalidateEnumerator(
+            IEnumerable enumerable);
+
+        private void RepeatTest(
+            Action<IEnumerator, object[], int> testCode, int iters = 3)
+        {
+            object[] items = GenerateItems(32);
+            IEnumerable enumerable = GetEnumerable(items);
+            IEnumerator enumerator = enumerable.GetEnumerator();
+            for (int i = 0; i < iters; i++)
+            {
+                testCode(enumerator, items, i);
+                if (IsResetNotSupported)
+                    enumerator = enumerable.GetEnumerator();
+                else
+                    enumerator.Reset();
+            }
+        }
+
+        private void RepeatTest(Action<IEnumerator, object[]> testCode, int iters = 3)
+        {
+            RepeatTest((e, i, it) => testCode(e, i), iters);
+        }
+
+        [Fact]
+        public void MoveNextHitsAllItems()
+        {
+            RepeatTest((enumerator, items) =>
+                       {
+                           var iterations = 0;
+                           while (enumerator.MoveNext())
+                           {
+                               iterations++;
+                           }
+                           Assert.Equal(items.Length, iterations);
+                       });
+        }
+
+        [Fact]
+        public void CurrentThrowsAfterEndOfCollection()
+        {
+            if (IsGenericCompatibility)
+                return; // apparently it is okay if enumerator.Current doesn't throw when the collection is generic.
+
+            RepeatTest((enumerator, items) =>
+                       {
+                           while (enumerator.MoveNext())
+                           {
+                           }
+                           Assert.Throws<InvalidOperationException>(
+                               () => enumerator.Current);
+                       });
+        }
+
+        [Fact]
+        public void MoveNextFalseAfterEndOfCollection()
+        {
+            RepeatTest((enumerator, items) =>
+                       {
+                           while (enumerator.MoveNext())
+                           {
+                           }
+
+                           Assert.False(enumerator.MoveNext());
+                       });
+        }
+
+        [Fact]
+        public void Current()
+        {
+            RepeatTest((enumerator, items, iteration) =>
```

Comment:
I can't really tell what this test is doing.  Can you make the name more descriptive, or add some comments?